[SPARK-30773][ML]Support NativeBlas for level-1 routines

yma11 opened the pull request:
### What changes were proposed in this pull request?
Change BLAS for part of level-1 routines(axpy, dot, scal(double, denseVector)) from java implementation to NativeBLAS when vector size>256

### Why are the changes needed?
In current ML BLAS.scala, all level-1 routines are fixed to use java
implementation. But NativeBLAS(intel MKL, OpenBLAS) can bring up to 11X
performance improvement based on performance test which apply direct
calls against these methods. We should provide a way to allow user take
advantage of NativeBLAS for level-1 routines. Here we do it through
switching to NativeBLAS for these methods from f2jBLAS.

### Does this PR introduce any user-facing change?
 Yes, methods axpy, dot, scal in level-1 routines will switch to NativeBLAS when it has more than nativeL1Threshold(fixed value 256) elements and will fallback to f2jBLAS if native BLAS is not properly configured in system.

### How was this patch tested?
Perf test direct calls level-1 routines

yma11 committed:
[SPARK-30773][ML]Support NativeBlas for level-1 routines

What changes were proposed in this pull request?
Change BLAS for part of level-1 routines(axpy, dot, scal) from java implementation to NativeBLAS

Why are the changes needed?
In current ML BLAS.scala, all level-1 routines are fixed to use java
implementation. But NativeBLAS(intel MKL, OpenBLAS) can bring up to 11X
performance improvement based on performance test which apply direct
calls against these methods. We should provide a way to allow user take
advantage of NativeBLAS for level-1 routines. Here we do it through
switching to NativeBLAS for these methods from f2jBLAS.

Does this PR introduce any user-facing change?
Yes, methods axpy, dot, scal in level-1 routines will switch to NativeBLAS when it has more than 256 elements and will fallback to
f2jBLAS if native BLAS is not properly configured in system.

How was this patch tested?
Perf test direct calls level-1 routines

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
I think the general problem is that Level 1 routines aren't faster. Are you sure it actually helps Spark? where and by how much? CC @mengxr

mengxr commented:
Could you provide reproducible microbenchmark results for both OpenBLAS and MKL on arrays of different sizes? Last time we tested it it didn't help. But that was years ago.

Note that JVM also improved SIMD. We should also check whether we can improve along that direction.

yma11 commented:
Hi @srowen and @mengxr ,
Thanks for your comments on this PR. Please refer the attached performance result. It includes the performance data using different BLAS implementation (f2jBLAS, Intel MKL and OpenBLAS). You can see nativeBLAS has quite big advantage in axpy(), dot() and scal(doule, dense) for both small and big vectors. As I tested, this advantage may can be improved more if do further tuning against the nativeBLAS variable such as MKL_NUM_THREADS/OPENBLAS_NUM_THREADS based on the data size, cores used, etc.
I also attached the microbenchmark lib for you to reproduce the result. Here is my test details for your information:

**Spark version:** 
master branch with head commit “e2d984aa1c79eb389cc8d333f656196b17af1c32: [SPARK-30733][R][HOTFIX] Fix SparkR tests per testthat and R version upgrade, and disable CRAN”
**Test environment:**
CPU: Skylake 6252(Intel(R) Xeon(R) Gold 6252 CPU @ 2.10GHz)
OS: CentOS 7
Kernel: 3.10.0-862.el7.x86_64
Java version: 1.8.0_112
MKL version: 2019
OpenBLAS version: libopenblas_haswellp-r0.3.8.dev
**Test command:**
/opt/spark/bin/spark-submit --name test  --master local --num-executors 1 --executor-cores 4 --executor-memory 10g --class org.intel.spark.TestMlAxpy(TestMlDot/TestMlScal) microbenchmark.jar

[Level1-routines-perf.xlsx](https://github.com/apache/spark/files/4204686/Level1-routines-perf.xlsx)
[microbenchmark.zip](https://github.com/apache/spark/files/4204689/microbenchmark.zip)

srowen commented:
These are microbenchmarks only, and for large vector sizes. The problem is that I don't think there's much if any win at smaller sizes, and that's often what the usage in Spark is like. Like, what's the difference for a 10-element vector?

mengxr commented:
Please also limit the blas threads to 1 because one core per task is the common setting.

yma11 commented:
Hi @srowen and @mengxr ,
With MKL_NUM_THREADS/OPENBLAS_NUM_THREADS set 1, I have tested 10-element, 100-element, 1000-element vectors.  Within 10-element vector, nativeBLAS has big gap compared with java implementation, which is quite easy to understand. But when elements number increases from 10 to 100, the performance gap drops dramatically and when the vectors have 1000 elements, nativeBLAS has quite big advantage against f2jBLAS in some dense vector operations. Please see details in the attached excel.
[Level1-routines-perf - part-2.xlsx](https://github.com/apache/spark/files/4212090/Level1-routines-perf.-.part-2.xlsx)

srowen commented:
Yeah, the problem is that it could be a win in some cases, but harmful in others. It depends on what real-world Spark usage is like, and many of these ops are used for small arrays in general.

It would be more persuasive to find what actual Spark usages benefit, and when, and by how much, from this change. It's possible some more targeted replacements seem to help much more than they may hurt.

mengxr commented:
@yma11 Thanks for re-running the benchmark! Could you confirm whether AVX is used by MKL/OpenBLAS? I think that is why it is faster on level-1 on large vectors.

It seems reasonable to switch between f2jBLAS and nativeBLAS based on vector size (256?). 

There is still one scenario we need to check. I don't think users will remember limiting native BLAS threads to 1. They might have 8 concurrent tasks running on a 8-core worker while each using multi-thread BLAS. I saw cases before that it significantly degrade the performance. It might be worth running a benchmark for this scenario.

yma11 commented:
Hi @srowen and @mengxr, 
I tested vector size 256  and nativeBLAS shows obvious perf gain compared with f2jBLAS in methods axpy(~1.7X), dot(~2.8X) and scal(double, dense)(>1.5X).  For MKL, I can confirm it uses AVX in the methods from output but for OpenBLAS, seems it haven't used AVX in level-1 routines as based on the info from [https://github.com/xianyi/OpenBLAS/blob/develop/README.md](url).
As to the MKL_NUM_THREADS and OPENBLAS_NUM_THREADS, limiting the threads to 1 doesn't always mean the best end-to-end performance. We used to test Kmeans using HiBench in a 1+4 cluster with MKL configured. Set the threads number to 1 or use default setting have no obvious performance change. By the way, using default MKL threads setting, MKL will bring 1.09X perf gain than java implementation in end-to-end.
I also have updated this PR to revert back to use java implementation for scal(sparse) and dspmv(). Please take a further review.

srowen commented:
In your k-means test, are the executors fairly fully utilized? like with 8 cores, do they have 8 tasks? that's the situation in which MKL_NUM_THREADS might make a difference, as trying to use 8*8=64 threads might hurt.

I can imagine just changing the logic in level 1 methods to use native if the array is, say, >= 256 elements. How about that? this should be a win.

yma11 commented:
@srowen In my kmeans test, each executor is configured with 6 cores and I can confirm it's fully utilized through web UI.  and under this condition, using default setting of MKL_NUM_THREADS is better than setting value to 1 but similar with setting value to 4. So I suggest to tune this env variable based on specific configuration of spark workloads.
To determine the threshold for logic change in level-1 methods, I have collected perf data for more vector sizes.  Thus based on the data of vector size(10, 100, 128, 150, 256, 1000), seems 128 is the point that nativeBlas has equal performance with java implementation and after that it shows obvious gain. Will it be reasonable to use this value as the logic change bar?
Attached includes perf data for 128/150 elements vectors.
[Level1-routines-perf - part-3.xlsx](https://github.com/apache/spark/files/4249399/Level1-routines-perf.-.part-3.xlsx)

srowen commented:
I think we could be conservative and let it switch over at operation on >= 256 elements? that seems like a win. It sounds like even when the cores are over-subscribed it could still be fast-er enough to justify than f2jblas

winningsix commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -27,15 +27,35 @@ private[spark] object BLAS extends Serializable {
 
   @transient private var _f2jBLAS: NetlibBLAS = _
   @transient private var _nativeBLAS: NetlibBLAS = _
+  @transient private val vectorSizeThreshold: Int = 256
```
Hint: Make this final and static

yma11 commented:
Hi @srowen and @mengxr, I have updated this PR using 256 as the value for nativaBLAS and java implementation switch.  Pls take a further view. Thanks.

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -27,15 +27,35 @@ private[spark] object BLAS extends Serializable {
 
   @transient private var _f2jBLAS: NetlibBLAS = _
   @transient private var _nativeBLAS: NetlibBLAS = _
+  @transient private val vectorSizeThreshold: Int = 256
```
This does not need to be transient. It's an int.

kiszk replied:
`256` depends on runtime environment such as CPU spec, JVM version, and so on.   
It should be taken from the property.

yma11 replied:
@srowen I think @kiszk make a good point. With properly tuned based on specific workload and runtime environment, nativeBLAS may can bring much benefit than f2jBLAS even vector size < 256.  Making it configurable can provide user a simpler way to use nativeBLAS in Spark ML when vector size is not so big. Do you think so? if so, we need to add spark-core dependency in spark-mllib-local package.
Thanks @kiszk for your comments.

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -27,15 +27,35 @@ private[spark] object BLAS extends Serializable {
+  // vector size >= 256, otherwise use f2jBLAS instead
+  private[ml] def getBLAS(vectorSize: Int): NetlibBLAS = {
+    if (vectorSize < vectorSizeThreshold) {
+      return f2jBLAS
```
You can omit "return"

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -27,15 +27,35 @@ private[spark] object BLAS extends Serializable {
+    _nativeBLAS
+  }
+
+  // For level-1 functions scal(double, SparseVector) and dspmv, use f2jBLAS for better performance.
```
I think you can remove this comment

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -27,15 +27,35 @@ private[spark] object BLAS extends Serializable {
-  // For level-1 routines, we use Java implementation.
+  // For level-3 routines, we use the native BLAS.
+  // if native BLAS is not properly configured in system, will automatically fallback to f2jBLAS
+  private[ml] def nativeBLAS: NetlibBLAS = {
```
I don't think you need to move this or expand its visibility?

yma11 replied:
I add a note in ml-guide to highlight nativeBLAS usage depends on proper configuration in system otherwise will fallback to java implementation.

srowen replied:
This still can move back to where it was

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -123,7 +143,7 @@ private[spark] object BLAS extends Serializable {
   private def dot(x: DenseVector, y: DenseVector): Double = {
     val n = x.size
-    f2jBLAS.ddot(n, x.values, 1, y.values, 1)
+    getBLAS(x.size).ddot(n, x.values, 1, y.values, 1)
```
just use n

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -220,20 +240,12 @@ private[spark] object BLAS extends Serializable {
       case sx: SparseVector =>
         f2jBLAS.dscal(sx.values.length, a, sx.values, 1)
```
I think you can apply the optimization here too; just depends on the length of sx.values

yma11 replied:
done.

kiszk reviewed (changes requested):
`256` should be configurable.

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -218,22 +234,14 @@ private[spark] object BLAS extends Serializable {
     x match {
       case sx: SparseVector =>
-        f2jBLAS.dscal(sx.values.length, a, sx.values, 1)
+        getBLAS(sx.size).dscal(sx.values.length, a, sx.values, 1)
```
Not .size, but .values.length here. The actual array is smaller than the size, usually.

yma11 replied:
Resolved in commit https://github.com/apache/spark/pull/27546/commits/118a0d8364641197ccf6e141ae5942309cd27909.

srowen reviewed (commented):
I don't know if it needs to be configurable, but it's not out of the question.

kiszk commented:
This performance experiment is done only one environment (one CPU, one JVM, and others). Then, we say `256` looks good threshold.

I am afraid that other environments have other thresholds (e.g. 128, 1024, or others). This is why the threshold should be configurable.

srowen commented:
Yeah I was just thinking of keeping it conservative, where it should always be a win. OK add an .ml config somewhere, I suppose. Maybe `spark.blas.nativeL1Threshold`?

kiszk commented:
The prop name sound good to me

yma11 committed:
use actual array size as value to determine BLAS type for Sparse vector

yma11 committed:
make nativeL1Threshold configurable

yma11 committed:
update spark configuration file for new property 'spark.ml.blas.nativeL1Threshold'

srowen commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -1541,4 +1541,10 @@ package object config {
 
+  private[spark] val NATIVE_L1_THRESHOLD =
+    ConfigBuilder("spark.ml.blas.nativeL1Threshold")
+      .doc("vector size for logic change between f2jBLAS and natveBLAS in " +
```
"Minimum vector size to use native BLAS libraries (if available) for Level 1 routines rather than Java implementation"
I'd also actually remove the "ml" from the property name as it isn't in an ML package, and isn't specific to ML. However we can document it in the context of ML as that is its only real use right now.

yma11 commented:
> Yeah I was just thinking of keeping it conservative, where it should always be a win. OK add an .ml config somewhere, I suppose. Maybe `spark.blas.nativeL1Threshold`?

Hi @srowen @kiszk @mengxr,  I have add a new property in spark and also updated the configuration doc.  Please take a review.

srowen commented on mllib-local/pom.xml:
```diff
@@ -61,13 +61,17 @@
       This spark-tags test-dep is needed even though it isn't used in this module, otherwise testing-cmds that exclude
       them will yield errors.
     -->
+    <dependency>
```
This has to be removed; it can't depend on core.
I think that means you need to move the property out of core.

yma11 replied:
Without dependency of core, ConfigBuilder/ConfigEntry won't be accessible. So I'd propose to add an environment variable in spark-env.sh. see new commit for details.

srowen replied:
Ah, right. OK, in that case I would leave the property in `core`, but reference it only by name (unfortunately) here in `mllib-local`, with a comment about why. You can remove `.ml` from the name too; it's not specific to ML really.

I don't think env variables are a good solution here.

winningsix replied:
hmm, instead env variables, we can pass this configuration via blas property file considering ```mllib-local``` package does not have access to ```core``` module.

srowen replied:
A property file is even more convoluted. You can still access the property; you just can't use the config API or symbol.

yma11 replied:
If we leave the property in core, we need to import org.apache.spark.internal.config to reference it in mllib-local thus still can't avoid dependency of core module.  Another option is that we implement a logic to directly read this value from spark-defaults.conf through method like System.getProperties, which will need no modification and dependency of core module.

srowen replied:
You don't, you just need to reference the config key by name, without references to symbols in core. This is a fine compromise.

yma11 replied:
Do you mean get the value using SparkConf by key? But SparkConf also has dependency on core module.

srowen replied:
Ah, OK that is not even available, right. I think I'd just make it a hard-coded constant then. I don't think we can reasonably expect people to find and discover and env variable that probably isn't that meaningful to tune. Just set it conservatively. 512?

yma11 replied:
Hi @srowen, how about 256 which we have finalized in previous discussions?

srowen replied:
Sure, I'm just trying to think of a threshold that should easily be a win regardless of other factors.

yma11 replied:
revert configurable related commits and now use 256 as the nativeL1Threshold.

srowen commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/BLAS.scala:
```diff
@@ -20,22 +20,42 @@ package org.apache.spark.ml.linalg
   @transient private var _f2jBLAS: NetlibBLAS = _
   @transient private var _nativeBLAS: NetlibBLAS = _
+  private val nativeL1Threshold =
+    new SparkConf().getInt(NATIVE_L1_THRESHOLD.key, 256)
```
You should be able to access the value from the ConfigEntry directly, and not repeat the default

yma11 committed:
use environment variable to configure BLAS
a) add ml/package.scala in mllib-local and delete the old one in mllib
to avoid duplication
b) update configuration.md to add SPARK_BLAS_NATIVE_L1_THRESHOLD
c) make BLAS switch based on environment varibale SPARK_BLAS_NATIVE_L1_THRESHOLD

yma11 committed:
Revert "make nativeL1Threshold configurable"

This reverts commit 118a0d8364641197ccf6e141ae5942309cd27909.

zhengruifeng commented:
Should `mllib.linalg.BLAS` also be changed? Some algorithms now are implemented in the `.mllib` side.

yma11 committed:
use nativeL1Threshold as the variable name

yma11 commented:
> n the `.mllib` side.
mllib package is in maintenance mode that why change applied in ml.

zhengruifeng commented:
@yma11 Yes, but there are still some algorithms have not been migrated to `.ml`.

BTW, did you also change `.mllib.linalg.BLAS` in your KMeans tests? Since KMeans is actually implemented in the `.mllib` side, so `.mllib.linalg.BLAS` is used instead of `.ml.linalg.BLAS`.

yma11 commented:
> @yma11 Yes, but there are still some algorithms have not been migrated to `.ml`.
> 
> BTW, did you also change `.mllib.linalg.BLAS` in your KMeans tests? Since KMeans is actually implemented in the `.mllib` side, so `.mllib.linalg.BLAS` is used instead of `.ml.linalg.BLAS`.

Yes,  in our Kmeans test, we modified the BLAS.scala in mllib and now trying to apply the change in mllib-local since mllib in maintenance mode. This change actually can't be leveraged by mllib Kmeans directly and if you want to do so, you need change the import location of BLAS.axpy to ml instead of mllib.

srowen commented:
Ah yes we do want to apply this across the board to all BLAS support.

yma11 commented:
> Ah yes we do want to apply this across the board to all BLAS support.

So let's apply same change in mllib/BLAS?

srowen commented:
Yes that's useful.

yma11 committed:
apply nativeL1Threshold change to mllib BLAS

yma11 commented:
> Yes that's useful.

Update PR with new commit for mllib BLAS. Please check.

srowen reviewed (commented):
I think it's coherent. @mengxr @kiszk WDYT? the problem is we can't reasonably plumb through config to mllib-local. We could chose to make it configurable only on the mllib side, but not sure how much it matters. Strong feelings?

kiszk commented:
Umm, I believe that this value depends on the environment (CPU, lib. version, JIT compiler version, ...). According to the past experiments, in general, the constants will be replaced with the property later.   

How about following [this approach](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/feature/Word2Vec.scala#L696
) in mllib?

srowen commented:
Yeah, the point of the above discussion is unfortunately there is no SparkConf in mllib-local, so that's not possible. It is possible in mllib. We could make it configurable just for mllib?

mengxr commented:
I'm okay with a hardcoded constant. We are start with a safer one, say 512, if there are concerns around environment differences.

yma11 commented:
> I'm okay with a hardcoded constant. We are start with a safer one, say 512, if there are concerns around environment differences.

I believe 256 is a very safe threshold since based on our experiments on this value, nativeBLAS has quite obvious advantage compared with f2jBLAS(at least 1.5X in the methods), without detailed tuning on nativeBLAS. Make it configurable in mllib will provide direct and flxible help for workloads like Kmeans. With fixed threshold is also okay.

mengxr commented:
@yma11 You also need to try old generation CPUs / RAMs to confirm 256 is absolutely safe. We cannot assume all users are on the latest generation.

kiszk commented:
Regarding configuration, since `mllib` has [a dependency on `core`](https://github.com/apache/spark/blob/master/mllib/pom.xml#L43), it looks ok that `mllib-local` also has a dependency on `core` newly.   
Is it not fine?

srowen commented:
I think the point of mllib-local was to contain code that can be used outside of a Spark app, so it can't have that dependency.

srowen commented:
I think the two remaining issues are:
- Should this threshold be configurable, only within `mllib` at least?
- Should the default / hard-coded value be larger, to be 'safer'?

Possibly still partly configurable, yes; I don't feel strongly about it.
@mengxr 's last question was, how does this compare on older hardware, like Ivy Bridge or similar, maybe? I think we're aiming for a value that should not cause a slowdown on hardware people might be realistically running now, especially if they haven't tuned BLAS threads. That number might be higher than 256 and I expect it is, but just guessing.

yma11 commented:
> I think the two remaining issues are:
> 
> * Should this threshold be configurable, only within `mllib` at least?
> * Should the default / hard-coded value be larger, to be 'safer'?
> 
> Possibly still partly configurable, yes; I don't feel strongly about it.
> @mengxr 's last question was, how does this compare on older hardware, like Ivy Bridge or similar, maybe? I think we're aiming for a value that should not cause a slowdown on hardware people might be realistically running now, especially if they haven't tuned BLAS threads. That number might be higher than 256 and I expect it is, but just guessing.

I am trying to get perf data of an older IA platform to double check whether 256 should be a safe threshold, even it's not all the case but it can make us more confident. Not finished yet and will update you once get the result.

kiszk commented:
I see about configurable threshold. For now, it is fine with constant.  
It is great to measure thresholds on multiple environments.

yma11 commented:
Perf data for nativeBLAS(mkl, openblas) and java implementation got collected on IA platform **SandyBridge**, a quite old generation of CPU.  Generally, 256 is still a safe value under this scenario. Please refer attached for details.
[Level1-routines-perf - part-4.xlsx](https://github.com/apache/spark/files/4322257/Level1-routines-perf.-.part-4.xlsx)

srowen reviewed (approved):
I'm personally OK with a fixed threshold of 256 for now, if it's a win on Sandy Bridge.

yma11 commented:
@kiszk and @mengxr  any more comments? I can see there is an old change request from kiszk.

mengxr commented:
LGTM. Could you add some tests to cover both <256 and >=256 cases?

yma11 commented:
I have tested 128, 256 and 512 cases in Level1-routines-perf - part-4.xlsx.

kiszk commented:
Sorry for being late. LGTM.
I guess that @mengxr asked to add test cases into the Spark test cases or benchmark.

yma11 committed:
add unit test for nativeL1Threshold in ml/mllib BLAS

yma11 commented:
@mengxr simple ut added for nativeL1Threshold.

mengxr commented:
jenkins, test this please

SparkQA commented:
**[Test build #119873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119873/testReport)** for PR 27546 at commit [`abddf4c`](https://github.com/apache/spark/commit/abddf4c45e17f108ebf8f3c89300cfc444433b61).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24601/
Test PASSed.

SparkQA commented:
**[Test build #119873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119873/testReport)** for PR 27546 at commit [`abddf4c`](https://github.com/apache/spark/commit/abddf4c45e17f108ebf8f3c89300cfc444433b61).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119873/
Test PASSed.

srowen commented:
@mengxr @kiszk do you have an opinion on merging this to 3.0? I don't feel strongly. Might be nice to get it in as a potential perf improvement.

mengxr commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/BLASSuite.scala:
```diff
@@ -23,6 +23,15 @@ import org.apache.spark.ml.util.TestingUtils._
 
+  test("nativeL1Threshold") {
+    var vectorSize = 128
+    assert(getBLAS(vectorSize) == BLAS.f2jBLAS)
```
minor: you don't need `var`. just `assert(getBLAS(128) == BLAS.f2jBLAS)` is sufficient and cleaner.

yma11 replied:
UT updated to use constant directly.

mengxr reviewed (changes requested):
For the affected BLAS method, could you also add tests for size >= threshold?

mengxr commented:
I'm fine with merging into 3.0 too.

kiszk commented:
While it is fine to merge it master now, I am not positive to merge it into 3.0. This is because it looks nice to have. 

But, I do not disagree with merging to 3.0 if others agree with the merge.

yma11 committed:
update unit test to use constant directly

srowen commented:
Jenkins, test this please

SparkQA commented:
**[Test build #120005 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120005/testReport)** for PR 27546 at commit [`04559df`](https://github.com/apache/spark/commit/04559df1615b9d6a97cd7fed731ca6cccf55bb6a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24724/
Test PASSed.

SparkQA commented:
**[Test build #120005 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120005/testReport)** for PR 27546 at commit [`04559df`](https://github.com/apache/spark/commit/04559df1615b9d6a97cd7fed731ca6cccf55bb6a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120005/
Test PASSed.

yma11 commented:
> For the affected BLAS method, could you also add tests for size >= threshold?

I added tests for size 128, 256, 512, should cover what you want me to do, right?

srowen commented:
Merged to master/3.0
